src/dp_nat.c

```diff
@@ -296,6 +297,184 @@ void dp_nat_chg_ip(struct dp_flow *df, struct rte_ipv4_hdr *ipv4_hdr,
 	}
 }
 
+static void dp_calculate_icmp_checksum(struct rte_icmp_hdr *icmp_hdr, size_t icmp_len)
+{
+	uint32_t sum = 0;
+	uint16_t word;
+	uint8_t *ptr = (uint8_t *)icmp_hdr;
+
+	icmp_hdr->icmp_cksum = 0;
+
+	for (size_t i = 0; i < icmp_len; i += 2) {
+		word = (uint16_t)((uint16_t)ptr[i] | ((uint16_t)ptr[i + 1] << 8));
+		sum += word;
+	}
+
+	while (sum >> 16)
+		sum = (sum & 0xFFFF) + (sum >> 16);
+
+	icmp_hdr->icmp_cksum = ~((uint16_t)sum);
+}
+
+int dp_nat_chg_ipv6_to_ipv4_hdr(struct dp_flow *df, struct rte_mbuf *m, uint32_t nat_ip, rte_be32_t *dest_ip /* out */)
+{
+	struct rte_ether_hdr *eth_hdr;
+	struct rte_ipv6_hdr *ipv6_hdr;
+	struct rte_ipv4_hdr *ipv4_hdr;
+	struct rte_icmp_hdr *icmp_hdr;
+	struct rte_udp_hdr *udp_hdr;
+	struct rte_tcp_hdr *tcp_hdr;
+	uint8_t l4_proto;
+
+	eth_hdr = rte_pktmbuf_mtod(m, struct rte_ether_hdr *);
+	eth_hdr->ether_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+
+	ipv6_hdr = dp_get_ipv6_hdr(m);
+	*dest_ip = *(int *)&ipv6_hdr->dst_addr[12];
+	l4_proto = ipv6_hdr->proto;
+
+	// Adjust the packet data to fit IPv4
+	if (rte_pktmbuf_adj(m, sizeof(struct rte_ipv6_hdr) - sizeof(struct rte_ipv4_hdr)) == NULL)
+		return DP_ERROR;
+
+	// Move the Ethernet header to just before the IPv4 header
+	memmove(rte_pktmbuf_mtod(m, uint8_t *), eth_hdr, sizeof(struct rte_ether_hdr));
```

Comment:
I actually never thought of this in virtual services (where this function is also being performed).
It makes sense that it works, looking at the DPDK source code, but it does seem strange to "shorten" memory and then access it. Maybe a comment about it being intentional would help?

Reply:
Added a comment.